sql: TestMemoryLimit failed

sql.TestMemoryLimit [failed](https://tanzanite.cluster.engflow.com/invocation/63648dc0-5144-406e-b8ad-d6fd3ce448ab) on master @ [e747f6e6857a19d6048cb184b0f55c52cb8a6390](https://github.com/cockroachdb/cockroach/commits/e747f6e6857a19d6048cb184b0f55c52cb8a6390):


```
=== RUN   TestMemoryLimit
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestMemoryLimit2436186933
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    test_server_shim.go:294: migration-manager-find-jobs: flow 49dfd1b3: memory budget exceeded: 2713600 bytes requested, 2764952 currently allocated, 5242880 bytes in budget
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestMemoryLimit2436186933
--- FAIL: TestMemoryLimit (0.39s)
```

Parameters:
 - <code>=</code>
 - <code>attempt=1</code>
 - <code>run=12</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMemoryLimit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34322

--- a/pkg/sql/mem_limit_test.go
+++ b/pkg/sql/mem_limit_test.go
@@ -41,17 +41,23 @@ func TestMemoryLimit(t *testing.T) {
 	// BatchResponses do hit the memory limit.
 	//
 	// When the Streamer API is not used, then SQL creates a single BatchRequest
-	// for both blobs without setting TargetBytes, so the storage layer is happy
-	// to get two blobs only to hit the memory limit. A concurrency of 1 is
-	// sufficient.
+	// with TargetBytes of 10MiB (controlled by
+	// rowinfra.defaultBatchBytesLimitProductionValue), so the storage layer is
+	// happy to get two blobs only to hit the memory limit. A concurrency of 1
+	// is sufficient.
 	//
 	// When the Streamer API is used, TargetBytes are set, so each blob gets a
 	// separate BatchResponse. Thus, we need concurrency of 2 so that the
 	// aggregate memory usage exceeds the memory limit. It's also likely that
 	// the error is encountered in the SQL layer when performing accounting for
 	// the read datums.
+	//
+	// The size here is constructed in such a manner that both rows are returned
+	// in a single ScanResponse, meaning that both rows together (including
+	// non-blob columns) don't exceed 10MiB.
+	const blobSize = 5<<20 - 1<<10 /* 5MiB - 1KiB */
 	serverArgs := base.TestServerArgs{
-		SQLMemoryPoolSize: 5 << 20, /* 5MiB */
+		SQLMemoryPoolSize: 2*blobSize - 1,
 	}
 	serverArgs.Knobs.SQLEvalContext = &eval.TestingKnobs{
 		// This test expects the default value of
@@ -63,9 +69,9 @@ func TestMemoryLimit(t *testing.T) {
 	defer s.Stopper().Stop(ctx)
 	_, err := db.Exec("CREATE TABLE foo (id INT PRIMARY KEY, attribute INT, blob TEXT, INDEX(attribute))")
 	require.NoError(t, err)
-	_, err = db.Exec("INSERT INTO foo SELECT 1, 10, repeat('a', 3000000)")
+	_, err = db.Exec(fmt.Sprintf("INSERT INTO foo SELECT 1, 10, repeat('a', %d)", blobSize))
 	require.NoError(t, err)
-	_, err = db.Exec("INSERT INTO foo SELECT 2, 10, repeat('a', 3000000)")
+	_, err = db.Exec(fmt.Sprintf("INSERT INTO foo SELECT 2, 10, repeat('a', %d)", blobSize))
 	require.NoError(t, err)
 
 	for _, tc := range []struct {

